bls-cert-verify/src/cert_verify.rs

```diff
@@ -0,0 +1,270 @@
+//! The BLS Cert Verify logic.
+
+use {
+    bitvec::prelude::{BitVec, Lsb0},
+    rayon::prelude::*,
+    solana_bls_signatures::{
+        pubkey::{Pubkey as BlsPubkey, PubkeyProjective, VerifiablePubkey},
+        signature::SignatureProjective,
+    },
+    solana_signer_store::{decode, DecodeError},
+    solana_votor_messages::consensus_message::{Certificate, CertificateType},
+    thiserror::Error,
+};
+
+#[derive(Debug, Error, PartialEq)]
+pub enum CertVerifyError {
+    #[error("Failed to decode bitmap {0:?}")]
+    BitmapDecodingFailed(DecodeError),
+
+    #[error("Failed to aggregate public keys")]
+    KeyAggregationFailed,
+
+    #[error("Failed to serialize original vote")]
+    SerializationFailed,
+
+    #[error("The signature doesn't match")]
+    SignatureVerificationFailed,
+
+    #[error("Base 3 encoding on unexpected cert {0:?}")]
+    Base3EncodingOnUnexpectedCert(CertificateType),
+}
+
+fn aggregate_keys_from_bitmap<F>(
+    bit_vec: &BitVec<u8, Lsb0>,
+    rank_to_pubkey: &F,
+) -> Option<PubkeyProjective>
+where
+    F: Fn(usize) -> Option<BlsPubkey> + Sync,
+{
+    let pubkeys: Result<Vec<_>, _> = bit_vec
+        .iter_ones()
+        .filter_map(rank_to_pubkey)
+        .map(PubkeyProjective::try_from)
+        .collect();
+    let pubkeys = pubkeys.ok()?;
+    PubkeyProjective::par_aggregate(pubkeys.par_iter()).ok()
+}
+
+pub fn verify_base2_certificate<F>(
+    cert_to_verify: &Certificate,
+    bit_vec: &BitVec<u8, Lsb0>,
+    rank_to_pubkey: &F,
+) -> Result<(), CertVerifyError>
+where
+    F: Fn(usize) -> Option<BlsPubkey> + Sync,
+{
+    let original_vote = cert_to_verify.cert_type.to_source_vote();
+
+    let Ok(signed_payload) = bincode::serialize(&original_vote) else {
+        return Err(CertVerifyError::SerializationFailed);
+    };
+
+    let Some(aggregate_bls_pubkey) = aggregate_keys_from_bitmap(bit_vec, rank_to_pubkey) else {
+        return Err(CertVerifyError::KeyAggregationFailed);
+    };
+
+    if let Ok(true) =
+        aggregate_bls_pubkey.verify_signature(&cert_to_verify.signature, &signed_payload)
+    {
+        Ok(())
+    } else {
+        Err(CertVerifyError::SignatureVerificationFailed)
+    }
```

Comment:
Lots of if-else / let else statements; can simplify:

```suggestion
    let original_vote = cert_to_verify.cert_type.to_source_vote();
    let signed_payload =
        bincode::serialize(&original_vote).map_err(|_| CertVerifyError::SerializationFailed)?;
    let aggregate_pubkey =
        aggregate_keys_from_bitmap(bit_vec, rank_to_pubkey).ok_or(CertVerifyError::KeyAggregationFailed)?;

    aggregate_pubkey
        .verify_signature(&cert_to_verify.signature, &signed_payload)
        .ok()
        .filter(|&valid| valid)
        .ok_or(CertVerifyError::SignatureVerificationFailed)
```

Reply:
Kept only the last sentence because I think that's clearer.